---
title: "ROC#4"
output: html_document
date: "2025-02-28"
---

# Music Choice

Studio Ghibili

# Load Packages 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

if (!require("dplyr")) {install.packages("dplyr"); require("dplyr")}
if (!require("Seurat")) {install.packages("Seurat"); require("Seurat")}
if (!require("patchwork")) {install.packages("patchwork"); require("patchwork")}
if (!require("cowplot")) {install.packages("cowplot"); require("cowplot")}
if (!require("ggpubr")) {install.packages("ggpubr"); require("ggpubr")}
if (!require("plotly")) {install.packages("plotly"); require("plotly")}
if (!require("knitr")) {install.packages("knitr"); require("knitr")}
if (!require("htmlwidgets")) {install.packages("htmlwidgets"); require("htmlwidgets")}
if (!require("here")) {install.packages("here"); require("here")}
if (!require("EnhancedVolcano")) {BiocManager::install('EnhancedVolcano'); require("EnhancedVolcano")} # volcano plot
if (!require("tidyverse")) {install.packages("tidyverse"); require("tidyverse")} # for titying up data
if (!require("RColorBrewer")) {install.packages("RColorBrewer"); require("RColorBrewer")} # for color brewer
if (!require("sctransform")) {install.packages("sctransform"); require("sctransform")} # for data normalization
if (!require("glmGamPoi")) {BiocManager::install('glmGamPoi'); require("glmGamPoi")} # for data normalization, sctransform
if (!require("openxlsx")) {install.packages("openxlsx"); require("openxlsx")} # to save .xlsx files
if (!require("SoupX")) {install.packages("SoupX"); require("SoupX")}
if (!require("DoubletFinder")) {BiocManager::install('DoubletFinder'); require("DoubletFinder")}

set.seed((12345))
here()

```

# Questions 

1)	Look at your answers for last week’s homework and make sure to save an .RDS file that labels the cells that you considered doublets (these labels should come from the doublet/mixed clusters).
2)	Analyze the “1” dataset (same as week 2 homework) with the current pipeline:
a.	Uploaded as .h5 files, run SoupX and Doublet Detector.
i.	Save an object that has both singlets and doublets
ii.	Save an object that is only the singlets 
b.	Make a table that compares how many doublets the current pipeline identifies and how that relates to the populations of doublets you identified last week.
i.	What percentage of doublets did the current pipeline identify from the old doublet population?
c.	Use the metadata from the previous week’s homework to graph where the cells that were previously labeled as doublets are localized to. 
i.	Compare where the cells go between the object that has “singlets and doublets” versus only “singlets”. 
1.	Where did the	 doublet cells go?
3)	Supercluster and identify the cell types by adding 2 levels of metadata.
4)	Create a modular score that includes at least 5 genes that define mesangial cells
a.	Is there a cluster that contains mesangial cells? 
b.	Is it specific to mesangial cells…or are there more cell types there?
5)	Write a forloop that graphs a vlnplot for the top 3 genes that define each cluster
6)	Perform DEG analysis on the clusters of the saved object and save an Rdata file and an Excel Spreadsheet.
a.	Bonus points if you write the code to do it based on cell-type names and not cluster numbers.

# Question 1

Saved RDS file of HW #3 as Dataset2 onto the Week 3 test folder

# Question 2 


## Run SoupX and Doublet Detector

Soup X 
This Algorithm was developed by former medical student Jeremiah Reyes

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}
options(future.globals.maxSize = 1000 * 1024^2)
tod = Read10X_h5(here("1_raw_feature_bc_matrix.h5")) #Change
toc = Read10X_h5(here("1_filtered_feature_bc_matrix.h5")) #Change 
sc = SoupChannel(tod,toc)

#Make the Seurat object from the filtered control data
SO <- Read10X_h5(here("1_filtered_feature_bc_matrix.h5")) #Change 
SO <- CreateSeuratObject(counts = SO, project = "Peri-INTACT")  #Change

#Cluster the cells with Seurat
SO <- SCTransform(SO, verbose = F)
SO <- RunPCA(SO, verbose = F)
SO <- RunUMAP(SO, dims = 1:30, verbose = F)
SO <- FindNeighbors(SO, dims = 1:30, verbose = F)
SO <- FindClusters(SO, verbose = T)
 
meta <- SO@meta.data
umap <- SO@reductions$umap@cell.embeddings
clusters <- setNames(meta$seurat_clusters, rownames(meta))

#Sanity Check
length(clusters) #should be equal to nrow(sc$metaData)
nrow(sc$metaData)

sc <- setClusters(sc, clusters)
sc <- setDR(sc, umap)

#Estimate rho
sc = autoEstCont(sc)

#Clean the data
SO_out = adjustCounts(sc)
 
#Create a new Seurat Object out of the cleaned data
seurat.obj <- CreateSeuratObject(SO_out)

#Run this command if you get an error in total size etc..
#options(future.globals.maxSize = 1000 * 1024^2)

```
## Remove Doublets with the Doublet Detector Package

The output of the SoupX chunck feeds right into the doublet detector pipline which is based on [(McGinnis et al., 2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6853612/).  

```{r, echo=TRUE, error=FALSE, fig.align= 'center', message=FALSE, warning=FALSE}

VlnPlot(seurat.obj, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

# Minimal QC and Filtering (low quality cells) to let doublet find doublets

seurat.obj.f <- subset(seurat.obj, nFeature_RNA > 500)

VlnPlot(seurat.obj.f, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

seurat.obj.f

# Pre-process standard workflow
seurat.obj.f <- NormalizeData(object = seurat.obj.f)
seurat.obj.f <- FindVariableFeatures(object = seurat.obj.f)
seurat.obj.f <- ScaleData(object = seurat.obj.f)
seurat.obj.f <- RunPCA(object = seurat.obj.f)
ElbowPlot(seurat.obj.f, ndims = 40)

# PCs between 15-20
seurat.obj.f <- FindNeighbors(object = seurat.obj.f, dims = 1:30)
seurat.obj.f <- FindClusters(object = seurat.obj.f, resolution = 0.03)
seurat.obj.f <- RunUMAP(object = seurat.obj.f, dims = 1:30)
DimPlot(seurat.obj.f, reduction = "umap")

```

## paramSweep

A very time-consuming step.

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', results=F}

# Calculate each combination of pN and pK
sweep.res.list_seurat.obj.f <- paramSweep(seurat.obj.f, PCs = 1:20, sct = FALSE) 
```

## summarizeSweep

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

#Summarize each combination of pN and pK
sweep.stats_seurat.obj.f <- summarizeSweep(sweep.res.list_seurat.obj.f, GT = FALSE) 

#Select the pK that corresponds to max bcmvn to optimize doublet detection
bcmvn_seurat.obj.f <- find.pK(sweep.stats_seurat.obj.f)
pK <- bcmvn_seurat.obj.f %>% 
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 

#See pK in the Values Environment
pK <- as.numeric(as.character(pK[[1]]))

```

## summarizeSweep

```{r , echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}
 
# Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
annotations <- seurat.obj.f@meta.data$seurat_clusters  
 
homotypic.prop <- modelHomotypic(annotations)           
homotypic.prop
 
# 10X Multiplet Rate Table (the doublet ratio is # of cells recovered divided by 125000) https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-
 
nrow(seurat.obj.f@meta.data)

nExp_poi <- round(nrow(seurat.obj.f@meta.data) # To calculate cell number
                  /125000              # To calculate the doublet ratio
                  *nrow(seurat.obj.f@meta.data))
nExp_poi

nExp_poi_adj <- round(nExp_poi*(1-homotypic.prop))
 
```

## Question Ai
doubletFinder

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_doublets <- doubletFinder(seurat.obj.f,
                        PCs = 1:20,
                        pN = 0.25,
                        pK = pK,
                        nExp = nExp_poi_adj,
                        reuse.pANN = FALSE, sct = FALSE)
colnames(seurat.obj.f_doublets@meta.data)[6] <- "pANN"
colnames(seurat.obj.f_doublets@meta.data)[7] <- "DF.class"
head(seurat.obj.f_doublets@meta.data)
table(seurat.obj.f_doublets@meta.data$DF.class)

DimPlot(seurat.obj.f_doublets, group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nFeature_RNA", group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nCount_RNA", group.by = "DF.class")

```

## Question Aii
Subset Singlets 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- subset(seurat.obj.f_doublets, DF.class == "Singlet")
seurat.obj.f_singlets
DimPlot(seurat.obj.f_singlets, reduction = "umap")

```

### Remove Mitochondrial Genes

Because it isn't clear what the percentage of mitochondrial genes means in a single-*nucleus* RNAseq dataset, we will take out all the mitochondrial genes to remove thier effect on clustering. However, before we remove them from the dataset, we will calculate their values and *stash* them in the metadata. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- seurat.obj.f_singlets[!grepl("^mt-", rownames(seurat.obj.f_singlets)), ]

#Mito Sanity Check
counts <- GetAssayData(seurat.obj.f_singlets, assay = "RNA")
mito.genes <- grep(pattern = "^mt-", x = rownames(x = counts), value = TRUE) 
mito.genes #should be zero

DimPlot(seurat.obj.f_singlets, reduction = "umap", label = T)

```

### Initial Cluster Identification

This is a generic pipeline that I use to identify the cell clusters in an unbiased way. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', fig.width = 7, fig.height= 8}

seurat.obj.f_singlets.markers <- FindAllMarkers(seurat.obj.f_singlets, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5

DoHeatmap(seurat.obj.f_singlets, features = top5$gene) + NoLegend()

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 2, wt = avg_log2FC) -> top2

VlnPlot(seurat.obj.f_singlets, 
        features = unique(top2$gene),
        stack = TRUE, 
        flip = TRUE,
        pt.size = 0)+
        NoLegend()


#Cluster 0 - Slc5a12 = PT-S1
#Cluster 1 - Slc12a3 = DCT
#Cluster 2 - Unknown Gm 33472, Gm41162
#Cluster 3

```


### Dotplot

This dotplot code was writen by former PSI student Annie Lackey and has been a go-to for the color scheme for dot plots ever since. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

DotPlot(seurat.obj.f_singlets, features = unique(top2$gene), dot.scale = 8, cols = c("dodgerblue2", "coral2")) + RotatedAxis()

```

## Save Output Files

This is the code to save the seurat object to it keeps all of it's analysis information including clustering information. The most important part of the code here is to make sure to include a *.rds* at the end of the name. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

saveRDS(seurat.obj.f_singlets, here("Singlets.rds")) #Change
saveRDS(seurat.obj.f_doublets, here ("Doublets.rds"))

```
## Question B
Load in RDS File from previous ROC #3

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}


Dataset <- LoadSeuratRds("Dataset2.rds")

# Create a new metadata column in the Seurat object
Dataset@meta.data$DF.class <- ifelse(Dataset@meta.data$class %in% c("Doublet", "Triplet"), "Doublet", "Singlet")

# Verify the updated metadata
table(Dataset@meta.data$DF.class)

#Using this they now have same meta data class


```

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

table(Dataset@meta.data$class)

table(seurat.obj.f_doublets$DF.class)
```

My Dataset using the old pipeline calculated `268+280` doublets and the new pipeline calculated 859 doublets

The difference in doublets is `859-268+280` more than mine. 

### What percentage of doublets did the current pipeline identify from the old doublet population

```{r}
(859/548)*100

```

There was 156% more doublets identified from the current pipeline compared to the old doublet population.

## Question C

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

# deleted environment but saved as an rds file to save space and make environment neater. 

Dataset <- LoadSeuratRds("Dataset2.rds")
Singlets <- LoadSeuratRds("Singlets.rds")

Doublets <- LoadSeuratRds("Doublets.rds")

```

```{r}


DimPlot(Singlets, reduction = "umap", label = T)
DimPlot(Doublets, reduction = "umap", label = T)
DimPlot(Dataset, reduction = "umap", label = T)

# I have 3 different seurat object, I've played around and tried a lot but my problem is merging the metadata. 

```

```{r}


#metadata <- Singlets@meta.data

#metadata

#metadata1 <- Dataset@meta.data

#metadata1

#metadata2 <- Doublets@meta.data

#metadata2

#Singlets <- AddMetaData(Dataset, metadata)

#head(Singlets@meta.data)

# this just completely replaced the data set



```

```{r echo=FALSE, error=FALSE, fig.align='center', message=FALSE, warning=FALSE}

DimPlot(Singlets, reduction = "umap") + ggtitle("Adjusted Pipeline") + theme(plot.title = element_text(hjust = 0.5))

markers.to.plot1 <- c("Lrp2",         # PT
                      "Slc5a12",     # PT-S1
                      "Slc13a3",     # PT-S2
                      "Slc16a9",     # PT-S3
                      "Havcr1",      # Injured PT
                      "Epha7",       # dTL
                      "Slc12a1",     # TAL
                      "Cldn10",      # TAL
                      "Cldn16",      # TAL
                      "Nos1",        # MD
                      "Slc12a3",     # DCT
                      "Pvalb",       # DCT1
                      "Slc8a1",      # DCT2, CNT
                      "Aqp2",        # PC
                      "Slc4a1",      # IC-A
                      "Slc26a4",     # IC-B
                      "Upk1b",       # Uro
                      "Ncam1",       # PEC
                      "Pdgfrb",      # Perivascular
                      "Piezo2",      # Mesangial
                      "Pdgfra",      # Fib
                      "Acta2",       # Mural
                      "Nphs1",       # Podo
                      "Kdr",         # Capillary Endo
                      "Lyve1",       # Lymph
                      "Ptprc",       # Immune
                      "Cd74",        # Macrophage
                      "Skap1"        # B/T Cells
)

DotPlot(Singlets,
        features = markers.to.plot1,
        dot.scale = 8,
        dot.min = 0,
        scale.max = 100,
        scale.min = 0,
        col.min = -2.5,
        col.max = 2.5)+
  coord_flip()

```

```{r echo=FALSE, error=FALSE, fig.align='center', message=FALSE, warning=FALSE}

DimPlot(, reduction = "umap") + ggtitle("Adjusted Pipeline") + theme(plot.title = element_text(hjust = 0.5))

markers.to.plot1 <- c("Lrp2",         # PT
                      "Slc5a12",     # PT-S1
                      "Slc13a3",     # PT-S2
                      "Slc16a9",     # PT-S3
                      "Havcr1",      # Injured PT
                      "Epha7",       # dTL
                      "Slc12a1",     # TAL
                      "Cldn10",      # TAL
                      "Cldn16",      # TAL
                      "Nos1",        # MD
                      "Slc12a3",     # DCT
                      "Pvalb",       # DCT1
                      "Slc8a1",      # DCT2, CNT
                      "Aqp2",        # PC
                      "Slc4a1",      # IC-A
                      "Slc26a4",     # IC-B
                      "Upk1b",       # Uro
                      "Ncam1",       # PEC
                      "Pdgfrb",      # Perivascular
                      "Piezo2",      # Mesangial
                      "Pdgfra",      # Fib
                      "Acta2",       # Mural
                      "Nphs1",       # Podo
                      "Kdr",         # Capillary Endo
                      "Lyve1",       # Lymph
                      "Ptprc",       # Immune
                      "Cd74",        # Macrophage
                      "Skap1"        # B/T Cells
)

DotPlot(Doublets,
        features = markers.to.plot1,
        dot.scale = 8,
        dot.min = 0,
        scale.max = 100,
        scale.min = 0,
        col.min = -2.5,
        col.max = 2.5)+
  coord_flip()

```


```{r echo=TRUE, error=FALSE, fig.align='center', message=FALSE, warning=FALSE}

# Assign class names based on clusters
Singlets@meta.data <- Singlets@meta.data %>%
  mutate(class = case_when(
    seurat_clusters == 0 ~ "PT",
    seurat_clusters == 1 ~ "CNT",
    seurat_clusters == 2 ~ "PT",
    seurat_clusters == 3 ~ "EC",
    seurat_clusters == 4 ~ "TAL",
    seurat_clusters == 5 ~ "dTL",
    seurat_clusters == 6 ~ "Fib/Peri",
    seurat_clusters == 7 ~ "IC",
    seurat_clusters == 8 ~ "PODO"
  ))

# this command makes a new meta data column called "class" (you can change it) with a repeat of the data from seurat clusters column

```

```{r}

# Now that I created a Column that can Compare , I think I can merge the data sets to see where they go?? mnot sure...

Idents(Singlets) <- Singlets@meta.data$class

DimPlot(Singlets, reduction = "umap", label = T)
DimPlot(Doublets, reduction = "umap", label = T)
DimPlot(Dataset, reduction = "umap", label = T)

```

```{r}

Dataset <- AddMetaData(object = Dataset, metadata = "Old", col.name = "Condition")
Singlets <- AddMetaData(object = Singlets, metadata = "New", col.name = "Condition")

```

## Merging and Normal scRNAseq Analysis Pipeline

It's possible to merge more than 2 objects together by concatenating the `y =` term to `y = c(SO2, SO3)`

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

SOmerged <- merge(Dataset, y = c(Singlets), add.cell.ids = c("Old", "New"), project = "Kidney")

#check on metadata and level it.

head(SOmerged@meta.data)

SOmerged@meta.data$Condition <- factor(SOmerged@meta.data$Condition, levels = c("New", "Old"))

SOmerged <- NormalizeData(SOmerged, normalization.method = "LogNormalize", scale.factor = 10000)
SOmerged <- FindVariableFeatures(SOmerged, selection.method = "vst", nfeatures = 2000)
SOmerged.genes <- rownames(SOmerged)
SOmerged <- ScaleData(SOmerged, features = SOmerged.genes)
SOmerged <- RunPCA(SOmerged, features = VariableFeatures(object = SOmerged))
SOmerged <- FindNeighbors(SOmerged, dims = 1:15)
SOmerged <- FindClusters(SOmerged, resolution = 0.1)
SOmerged <- RunUMAP(SOmerged, dims = 1:15)

DimPlot(SOmerged, reduction = "umap")
f1 = DimPlot(SOmerged, reduction = "umap", group.by = "Condition")
f2 = DimPlot(SOmerged, reduction = "umap", group.by = "class")

f1 + f2
```
```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', }

SOmerged.integrated <- IntegrateLayers(object = SOmerged, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca",
    verbose = FALSE)

# re-join layers after integration
SOmerged.integrated[["RNA"]] <- JoinLayers(SOmerged.integrated[["RNA"]])

SOmerged.integrated <- FindNeighbors(SOmerged.integrated, reduction = "integrated.cca", dims = 1:30)
SOmerged.integrated <- FindClusters(SOmerged.integrated, resolution = 1)
SOmerged.integrated <- RunUMAP(SOmerged.integrated, dims = 1:30, reduction = "integrated.cca")

DimPlot(SOmerged.integrated, reduction = "umap")
DimPlot(SOmerged.integrated, reduction = "umap", group.by = c("Condition"))

#I think this is wrong but my computer can't run this command due to Data being too large anyway. 

```


```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

#Modular Score??? Mesangial Cell 

SOmerged <- JoinLayers(SOmerged)

Mes_marker_gene_list <- list(c(
  "Pdgfrb", "Pdgfra", "Itga8", "Acta2", "Des",
  "Gata3", "Cnn1", "Tns1", "Myh11", "Tagln"
))

SOmerged <- AddModuleScore(object = SOmerged, features = Mes_marker_gene_list, name = "Mes_score")


#Check for score in the metadata

head(SOmerged@meta.data)

```

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

library(RColorBrewer)


FeaturePlot(SOmerged,
            features = c("Mes_score1")) &
  scale_colour_gradientn(colours = rev(brewer.pal(n = 9, name = "RdYlBu")))

VlnPlot(SOmerged, "Mes_score1")


```


```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

DimPlot(SOmerged, reduction = "umap") + ggtitle("Adjusted Pipeline") + theme(plot.title = element_text(hjust = 0.5))


 #"Pdgfrb", "Pdgfra", "Itga8", "Acta2", "Des",
 # "Gata3", "Cnn1", "Tns1", "Myh11", "Tagln"

markers.to.plot2 <- c(
  "Pdgfrb",   # PT
  "Pdgfra",   # PT-S1
  "Itga8",    # PT-S2
  "Acta2",    # PT-S3
  "Des",      # Injured PT
  "Gata3",    # dTL
  "Cnn1",     # TAL
  "Tns1",     # TAL
  "Myh11",    # TAL
  "Tagln",     # MD
  "Piezo2"
)


DotPlot(SOmerged,
        features = markers.to.plot2,
        dot.scale = 8,
        dot.min = 0,
        scale.max = 100,
        scale.min = 0,
        col.min = -2.5,
        col.max = 2.5)+
  coord_flip()

```


Im assuming Cluster 7 contains most of the mesangial cells.

I forgot the website for the gene expression. (Lee alt 2015) is the only thing I remember


```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

list <- c("Pdgfrb", "Pdgfra", "Cfh")
names(list) <- list

for(i in list){
  A <- VlnPlot(seurat.obj.f_singlets, i) + ggtitle(paste("Pericyte Marker", i, sep=" "))
  print(A)
}

```



I believe I did some of these completely wrong so im going to restart again