---
title: "snRNAseq Analysis Week 3 Materials"
author: "Jonathan Nelson"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    toc: yes
    toc_depth: 4
    toc_float:
      collapsed: true
      number_sections: yes
    theme: journal
    df_print: paged
    code_folding: hide
    highlight: pygments
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Learning Objectives

At the end of Week 3, you will be able to: 

  1) Merge and integrate two datasets together 
  3) Remove Ambient RNA with the SoupX package
  4) Remove doublets with the doublet detector package
  5) Create modular scores to look for patterns of changes with groups of genes
  6) Exporting metadata from a Seurat object
  7) Merging metadata from a Seurat object
  8) Write a forloop <br>
      a. Write a forloop for graphing genes individually with unique labels<br>
      b. Write a forloop to store DEG lists<br>
      c. Move DEG lists into new environment
  9) Save DEG lists as an Rdata file
  10) Save DEG lists as an excel file
    
## Load Required Packages

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

if (!require("dplyr")) {install.packages("dplyr"); require("dplyr")}
if (!require("Seurat")) {install.packages("Seurat"); require("Seurat")}
if (!require("patchwork")) {install.packages("patchwork"); require("patchwork")}
if (!require("cowplot")) {install.packages("cowplot"); require("cowplot")}
if (!require("ggpubr")) {install.packages("ggpubr"); require("ggpubr")}
if (!require("plotly")) {install.packages("plotly"); require("plotly")}
if (!require("knitr")) {install.packages("knitr"); require("knitr")}
if (!require("htmlwidgets")) {install.packages("htmlwidgets"); require("htmlwidgets")}
if (!require("here")) {install.packages("here"); require("here")}
if (!require("EnhancedVolcano")) {BiocManager::install('EnhancedVolcano'); require("EnhancedVolcano")} # volcano plot
if (!require("tidyverse")) {install.packages("tidyverse"); require("tidyverse")} # for titying up data
if (!require("RColorBrewer")) {install.packages("RColorBrewer"); require("RColorBrewer")} # for color brewer
if (!require("sctransform")) {install.packages("sctransform"); require("sctransform")} # for data normalization
if (!require("glmGamPoi")) {BiocManager::install('glmGamPoi'); require("glmGamPoi")} # for data normalization, sctransform
if (!require("openxlsx")) {install.packages("openxlsx"); require("openxlsx")} # to save .xlsx files
if (!require("SoupX")) {install.packages("SoupX"); require("SoupX")}
if (!require("DoubletFinder")) {BiocManager::install('DoubletFinder'); require("DoubletFinder")}

set.seed((12345))
here()

```

# Merge and Integrate Two Datasets Together

Often times, the point of a single-cell RNAseq experiment is to compare groups together. 

Due to variability in experimental conditions, often times simply `merging` a dataset isn't enough (as we will see in this example) and we need to perform additional calculations to identify genes that can be used as *anchors* to group cells together. 

## Uploading and Analyzing Two snRNAseq Datasets

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

SO1 <- Read10X_h5(here("MH1_filtered_feature_bc_matrix.h5"))
SO2 <- Read10X_h5(here("MH3_filtered_feature_bc_matrix.h5"))

SO1 <- CreateSeuratObject(counts = SO1, project = "Dataset", min.cells = 3, min.features = 200)
SO2 <- CreateSeuratObject(counts = SO2, project = "Dataset", min.cells = 3, min.features = 200)

```

### Add meta data to each object to keep track of them.

In this case, SO1 is *Sham* and SO2 is treated with *CACPR*. I'm adding this information to the same metadata column that is called *Condition*

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

SO1 <- AddMetaData(object = SO1, metadata = "CACPR", col.name = "Condition")
SO2 <- AddMetaData(object = SO2, metadata = "Sham", col.name = "Condition")

```

## Merging and Normal scRNAseq Analysis Pipeline

It's possible to merge more than 2 objects together by concatenating the `y =` term to `y = c(SO2, SO3)`

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

SOmerged <- merge(SO1, y = c(SO2), add.cell.ids = c("CACPR", "Sham"), project = "Kidney")

#check on metadata and level it.

head(SOmerged@meta.data)

SOmerged@meta.data$Condition <- factor(SOmerged@meta.data$Condition, levels = c("Sham", "CACPR"))

SOmerged <- NormalizeData(SOmerged, normalization.method = "LogNormalize", scale.factor = 10000)
SOmerged <- FindVariableFeatures(SOmerged, selection.method = "vst", nfeatures = 2000)
SOmerged.genes <- rownames(SOmerged)
SOmerged <- ScaleData(SOmerged, features = SOmerged.genes)
SOmerged <- RunPCA(SOmerged, features = VariableFeatures(object = SOmerged))
SOmerged <- FindNeighbors(SOmerged, dims = 1:15)
SOmerged <- FindClusters(SOmerged, resolution = 0.1)
SOmerged <- RunUMAP(SOmerged, dims = 1:15)

DimPlot(SOmerged, reduction = "umap")
DimPlot(SOmerged, reduction = "umap", group.by = "Condition")

```
Notice that there is almost a complete separation between *Sham* and *CACPR* samples. This is probably the combination of a strong

To find genes that are shared between these two samples that can be used as anchors to cluster similar cells between the two sample we will use the *FindIntegrationAnchors* pipeline that was intiaily developed by former PSI high school student Annie Lackey which she learned from the Seurat website. 

https://satijalab.org/seurat/articles/integration_introduction.html

## FindIntegration Anchors Pipeline

In order to solve this problem we turn to a feature of Seurat that helps us integrate data. To prep the data we first have to split the object by treatment group. This creates the new feature `SOmerged.list`, so we can continue with standard normalization and finding the features of the new object.


### split the RNA measurements into two layers one for control cells, one for stimulated cells

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', }

SOmerged.integrated <- IntegrateLayers(object = SOmerged, method = CCAIntegration, orig.reduction = "pca", new.reduction = "integrated.cca",
    verbose = FALSE)

# re-join layers after integration
SOmerged.integrated[["RNA"]] <- JoinLayers(SOmerged.integrated[["RNA"]])

SOmerged.integrated <- FindNeighbors(SOmerged.integrated, reduction = "integrated.cca", dims = 1:30)
SOmerged.integrated <- FindClusters(SOmerged.integrated, resolution = 1)
SOmerged.integrated <- RunUMAP(SOmerged.integrated, dims = 1:30, reduction = "integrated.cca")

DimPlot(SOmerged.integrated, reduction = "umap")
DimPlot(SOmerged.integrated, reduction = "umap", group.by = c("Condition"))

```


### Comparison of Merged and Integrated Pipelines

Notice the addition of the ggplot2 code that is layered onto the `DimPlot` function in order to modify the appearance of the graph for clarity as `+ ggtitle("Merged")`.

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

DimPlot(SOmerged, reduction = "umap", group.by = "Condition") + ggtitle("Merged")

DimPlot(SOmerged.integrated, reduction = "umap", group.by = "Condition") + ggtitle("Integrated")

```

### Using Patchwork place the graphs side-by-side

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.height = 6, fig.width = 12, class.source = "fold-show"}

f1 <- DimPlot(SOmerged, reduction = "umap", group.by = "Condition") + ggtitle("Merged")

f2 <- DimPlot(SOmerged.integrated, reduction = "umap", group.by = "Condition") + ggtitle("Integrated")

f1 + f2

rm(SOmerged, SOmerged.integrated)

```

# Removing Ambient RNA with SoupX

The current strategy that we use to remove doublets relies on [(Young and Behjati, 2020)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7763177/)

In order for this code to work you need both the *raw_feature_bc_matrix* in addition to the *filteredfiltered_feature_bc_matrix*. 

This pipeline was written by former visiting medical student [(Jeremiah Reyes)](https://twitter.com/imyourNephBro). 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

tod = Read10X_h5(here("1_raw_feature_bc_matrix.h5")) #Change
toc = Read10X_h5(here("1_filtered_feature_bc_matrix.h5")) #Change 
sc = SoupChannel(tod,toc)

#Make the Seurat object from the filtered control data
SO <- Read10X_h5(here("1_filtered_feature_bc_matrix.h5")) #Change 
SO <- CreateSeuratObject(counts = SO, project = "Peri-INTACT")  #Change

#Cluster the cells with Seurat
SO <- SCTransform(SO, verbose = F)
SO <- RunPCA(SO, verbose = F)
SO <- RunUMAP(SO, dims = 1:30, verbose = F)
SO <- FindNeighbors(SO, dims = 1:30, verbose = F)
SO <- FindClusters(SO, verbose = T)
 
meta <- SO@meta.data
umap <- SO@reductions$umap@cell.embeddings
clusters <- setNames(meta$seurat_clusters, rownames(meta))

#Sanity Check
length(clusters) #should be equal to nrow(sc$metaData)
nrow(sc$metaData)

sc <- setClusters(sc, clusters)
sc <- setDR(sc, umap)

#Estimate rho
sc = autoEstCont(sc)

#Clean the data
SO_out = adjustCounts(sc)
 
#Create a new Seurat Object out of the cleaned data
seurat.obj <- CreateSeuratObject(SO_out)
```

# Remove Doublets with the Doublet Detector Package

The output of the SoupX chunck feeds right into the doublet detector pipline which is based on [(McGinnis et al., 2019)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6853612/).  

```{r, echo=TRUE, error=FALSE, fig.align= 'center', message=FALSE, warning=FALSE}

VlnPlot(seurat.obj, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

# Minimal QC and Filtering (low quality cells) to let doublet find doublets

seurat.obj.f <- subset(seurat.obj, nFeature_RNA > 500)

VlnPlot(seurat.obj.f, features = c("nFeature_RNA", "nCount_RNA"), ncol = 2)

seurat.obj.f

# Pre-process standard workflow
seurat.obj.f <- NormalizeData(object = seurat.obj.f)
seurat.obj.f <- FindVariableFeatures(object = seurat.obj.f)
seurat.obj.f <- ScaleData(object = seurat.obj.f)
seurat.obj.f <- RunPCA(object = seurat.obj.f)
ElbowPlot(seurat.obj.f, ndims = 40)

# PCs between 15-20
seurat.obj.f <- FindNeighbors(object = seurat.obj.f, dims = 1:30)
seurat.obj.f <- FindClusters(object = seurat.obj.f, resolution = 0.03)
seurat.obj.f <- RunUMAP(object = seurat.obj.f, dims = 1:30)
DimPlot(seurat.obj.f, reduction = "umap")

```

## paramSweep

A very time-consuming step.

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', results=F}

# Calculate each combination of pN and pK
sweep.res.list_seurat.obj.f <- paramSweep(seurat.obj.f, PCs = 1:20, sct = FALSE) 
```

## summarizeSweep

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

#Summarize each combination of pN and pK
sweep.stats_seurat.obj.f <- summarizeSweep(sweep.res.list_seurat.obj.f, GT = FALSE) 

#Select the pK that corresponds to max bcmvn to optimize doublet detection
bcmvn_seurat.obj.f <- find.pK(sweep.stats_seurat.obj.f)
pK <- bcmvn_seurat.obj.f %>% 
  filter(BCmetric == max(BCmetric)) %>%
  select(pK) 

#See pK in the Values Environment
pK <- as.numeric(as.character(pK[[1]]))

```

## summarizeSweep

```{r , echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}
 
# Homotypic Doublet Proportion Estimate -------------------------------------------------------------------------------------
annotations <- seurat.obj.f@meta.data$seurat_clusters  
 
homotypic.prop <- modelHomotypic(annotations)           
homotypic.prop
 
# 10X Multiplet Rate Table (the doublet ratio is # of cells recovered divided by 125000) https://kb.10xgenomics.com/hc/en-us/articles/360001378811-What-is-the-maximum-number-of-cells-that-can-be-profiled-
 
nrow(seurat.obj.f@meta.data)

nExp_poi <- round(nrow(seurat.obj.f@meta.data) # To calculate cell number
                  /125000              # To calculate the doublet ratio
                  *nrow(seurat.obj.f@meta.data))
nExp_poi

nExp_poi_adj <- round(nExp_poi*(1-homotypic.prop))
 
```

## doubletFinder

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_doublets <- doubletFinder(seurat.obj.f,
                        PCs = 1:20,
                        pN = 0.25,
                        pK = pK,
                        nExp = nExp_poi_adj,
                        reuse.pANN = FALSE, sct = FALSE)
colnames(seurat.obj.f_doublets@meta.data)[6] <- "pANN"
colnames(seurat.obj.f_doublets@meta.data)[7] <- "DF.class"
head(seurat.obj.f_doublets@meta.data)
table(seurat.obj.f_doublets@meta.data$DF.class)

DimPlot(seurat.obj.f_doublets, group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nFeature_RNA", group.by = "DF.class")

VlnPlot(seurat.obj.f_doublets, "nCount_RNA", group.by = "DF.class")

```

## Subset Singlets

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- subset(seurat.obj.f_doublets, DF.class == "Singlet")
seurat.obj.f_singlets
DimPlot(seurat.obj.f_singlets, reduction = "umap")

```

### Remove Mitochondrial Genes

Because it isn't clear what the percentage of mitochondrial genes means in a single-*nucleus* RNAseq dataset, we will take out all the mitochondrial genes to remove thier effect on clustering. However, before we remove them from the dataset, we will calculate their values and *stash* them in the metadata. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

seurat.obj.f_singlets <- seurat.obj.f_singlets[!grepl("^mt-", rownames(seurat.obj.f_singlets)), ]

#Mito Sanity Check
counts <- GetAssayData(seurat.obj.f_singlets, assay = "RNA")
mito.genes <- grep(pattern = "^mt-", x = rownames(x = counts), value = TRUE) 
mito.genes #should be zero

DimPlot(seurat.obj.f_singlets, reduction = "umap", label = T)

```

### Initial Cluster Identification

This is a generic pipeline that I use to identify the cell clusters in an unbiased way. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', fig.width = 7, fig.height= 8}

seurat.obj.f_singlets.markers <- FindAllMarkers(seurat.obj.f_singlets, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5

DoHeatmap(seurat.obj.f_singlets, features = top5$gene) + NoLegend()

seurat.obj.f_singlets.markers %>%
    group_by(cluster) %>%
    top_n(n = 2, wt = avg_log2FC) -> top2

VlnPlot(seurat.obj.f_singlets, 
        features = unique(top2$gene),
        stack = TRUE, 
        flip = TRUE,
        pt.size = 0)+
        NoLegend()

```

### Dotplot

This dotplot code was writen by former PSI student Annie Lackey and has been a go-to for the color scheme for dot plots ever since. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

DotPlot(seurat.obj.f_singlets, features = unique(top2$gene), dot.scale = 8, cols = c("dodgerblue2", "coral2")) + RotatedAxis()

```

## Save Output Files

This is the code to save the seurat object to it keeps all of it's analysis information including clustering information. The most important part of the code here is to make sure to include a *.rds* at the end of the name. 

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

#saveRDS(seurat.obj.f_singlets, here("Singlets.rds)) #Change

```

# Creating a Modular Scores 

Sometimes instead of graphing individual genes, we would like to graph the expression level of a group of genes. 

For instance, we might want to graph a set of genes that defines a special function, such as *fibrosis* or *cell division*. In this case, we will graph a set of genes that defines 2 sub-segmenets of the DCT (DCT1 and DCT2).

*Note*: As a wierd oddity of how the genes are calculated and exported, they add a "1" in the name of the metadata column and so the column to call this will be `DCTscore1`

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}


DCT1_marker_gene_list <- list(c("Erbb4", "Egf", "Trpm7", "Fgf13", "Col5a2", "Umod", "Ptgfr", "Stk32b", "Rtl4", "Abca13")) # Top 10 genes from DCT-INTACT-NK DCT1 cluster
DCT2_marker_gene_list <- list(c("Slc8a1", "Arl15", "Calb1", "Slc2a9", "Phactr1", "Gls", "S100g", "Kl", "Klk1", "Egfem1")) # Top 10 genes from DCT-INTACT-NK DCT2 cluster

seurat.obj.f_singlets <- AddModuleScore(object = seurat.obj.f_singlets, features = DCT1_marker_gene_list, name = "DCT1_score")
seurat.obj.f_singlets <- AddModuleScore(object = seurat.obj.f_singlets, features = DCT2_marker_gene_list, name = "DCT2_score")

#Check for score in the metadata

head(seurat.obj.f_singlets@meta.data)

```

## Graph Modular Score as FeaturePlot or VlnPlot

For this code I use the `RColorBrewer`  package in order to graph a dynamic range of colors for the score. 

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center'}

FeaturePlot(seurat.obj.f_singlets,
            features = c("DCT1_score1", "DCT2_score1")) &
  scale_colour_gradientn(colours = rev(brewer.pal(n = 9, name = "RdYlBu")))

VlnPlot(seurat.obj.f_singlets, "DCT1_score1")
VlnPlot(seurat.obj.f_singlets, "DCT2_score1")

```

# Exporting metadata from a Seurat object

Using this code, the rownames become the barcodes, which is exactly where they need to be in order to be matched up with barcodes of another object.

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

metadata <- seurat.obj.f_singlets@meta.data

metadata

```

# Merging metadata from a dataframe in the envorment

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

seurat.obj.f_singlets <- AddMetaData(seurat.obj.f_singlets, metadata)

head(seurat.obj.f_singlets@meta.data)

```

# Writing a forloop

forloops are powerful coding terms that allow you to repeatedly perform the same chuck of code on multiple groups of data. 

The basic steps are to:

1) Define a list
2) Define the function that will cycle through that list

Note: By default, an r object isn't aware of it's name. You need to give the object in the environment a name using the `names` code in order to call on it later if you want to use it for making a label in a graph or giving using it as part of the file name.

For more information on writing a forloop, consider reviewing this excellent resource. 

https://www.dataquest.io/blog/for-loop-in-r/#

## Write a forloop for graphing genes individually with unique labels

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

list <- c("Pdgfrb", "Pdgfra", "Cfh")
names(list) <- list

for(i in list){
  A <- VlnPlot(seurat.obj.f_singlets, i) + ggtitle(paste("Pericyte Marker", i, sep=" "))
  print(A)
}

```

## Write a forloop to calculate DEGs from every cluster in a Seurat Object

This code is written to identify DEGs based on the cluster identifier by their number (not by other meta data). 

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

clusters <- unlist(levels(seurat.obj.f_singlets$seurat_clusters))
names(clusters) <- as.numeric(clusters)

DEG.env <- new.env()

ls(DEG.env)

for (i in clusters) {
  DEGs <- FindMarkers(seurat.obj.f_singlets, ident.1 = i, min.pct = 0.25, logfc.threshold = 0.5)
  DEGs <- DEGs[order(DEGs[,"avg_log2FC"], decreasing = TRUE),]
  DEGs <- rownames_to_column(DEGs, "Gene")
  assign(names(clusters[i]), DEGs, envir = DEG.env)
}

ls(DEG.env)

```

# Save DEG lists as an Rdata file

This code takes all of the objects in the `DEG.env` and saves them as an `.Rdata` file. This file can be incredibly useful for opening up the DEG list in R for further analysis. 

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

list_objects <- ls(DEG.env)
save(list = list_objects, file = here("Test.RData"), envir = DEG.env)

```

# Save DEG lists an excel file

Often, we want to communicate our science with people who aren't familiar with coding and R and excel makes for a useful medium to share DEG lists. This code will take all the DEG lists that you created and placed into the `DEG.env` and write them as an excel file with each tab named after the DEG list output. 

```{r echo=TRUE, warning=FALSE, error=FALSE, message=FALSE, fig.align = 'center', class.source = "fold-show"}

allobj <- ls(DEG.env)
alldf <- sapply(allobj, get, simplify = FALSE, envir = DEG.env)
alldf <- alldf[sapply(alldf, inherits, "data.frame")]
alldf <- alldf[order(as.numeric(names(alldf)))]

write.xlsx(alldf, sheetName = names(alldf), file = here("Test.xlsx"))

```

# Session Info

```{r, echo=TRUE, warning=FALSE, error=FALSE, message=FALSE}

sessionInfo()

```


